pkg/operator/target_status_test.go

```diff
@@ -0,0 +1,1532 @@
+			port := int32(19090)
+			prometheusTargetMap := make(map[string]*prometheusv1.TargetsResult, amount)
+			targetsExpected := make([]*prometheusv1.TargetsResult, 0, amount)
+			kubeClientBuilder := fake.NewClientBuilder().WithScheme(scheme).WithObjects(&appsv1.DaemonSet{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      NameCollector,
+					Namespace: opts.OperatorNamespace,
+				},
+				Spec: appsv1.DaemonSetSpec{
+					Selector: &metav1.LabelSelector{},
+					Template: corev1.PodTemplateSpec{
+						ObjectMeta: metav1.ObjectMeta{},
+						Spec: corev1.PodSpec{
+							Containers: []corev1.Container{{
+								Name: "prometheus",
+								Ports: []corev1.ContainerPort{{
+									Name:          "prom-metrics",
+									ContainerPort: port,
+								}},
+							}},
+						},
+					},
+				},
+			})
+			for i := 0; i < amount; i++ {
+				pod := &corev1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      fmt.Sprintf("pod-%d", i),
+						Namespace: opts.OperatorNamespace,
+					},
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{{
+							Name: "prometheus",
+						}},
+					},
+					Status: corev1.PodStatus{
+						Phase: corev1.PodRunning,
+						PodIP: fmt.Sprint(i),
+						ContainerStatuses: []corev1.ContainerStatus{{
+							Name:  "prometheus",
+							Ready: true,
+						}},
+					},
+				}
+				kubeClientBuilder.WithObjects(pod)
+
+				prometheusTargetMap[getPodKey(pod, port)] = &prometheusv1.TargetsResult{
+					Active: []prometheusv1.ActiveTarget{{
+						Health: "up",
+						Labels: map[model.LabelName]model.LabelValue{
+							"instance": model.LabelValue(fmt.Sprint(amount)),
+						},
+						ScrapePool:         "PodMonitoring/gmp-test/prom-example-1/metrics",
+						LastError:          "err x",
+						LastScrapeDuration: 1.2,
+					}},
+					Dropped: []prometheusv1.DroppedTarget{},
+				}
+
+				target := &prometheusv1.TargetsResult{
+					Active: []prometheusv1.ActiveTarget{{
+						Health:     "up",
+						LastError:  "err x",
+						ScrapePool: "PodMonitoring/gmp-test/prom-example-1/metrics",
+						Labels: model.LabelSet(map[model.LabelName]model.LabelValue{
+							"instance": model.LabelValue(fmt.Sprint(amount)),
+						}),
+						LastScrapeDuration: 1.2,
+					}},
+					Dropped: []prometheusv1.DroppedTarget{},
+				}
+				targetsExpected = append(targetsExpected, target)
+			}
+
+			targetFetcher := mockTargetFetcher{
+				prometheusTargets: prometheusTargetMap,
+			}
+			kubeClient := kubeClientBuilder.Build()
+
+			targets, err := fetchTargets(ctx, logger, opts, targetFetcher.getTarget, kubeClient)
+			if err != nil {
+				t.Fatal("Unable to fetch targets", err)
+			}
+
+			if len(targets) != amount {
+				t.Fatalf("Wrong target amount: %s", cmp.Diff(len(targets), amount))
+			}
+
+		Outer:
+			for targetIndex, targetExpected := range targetsExpected {
+				for _, foundTarget := range targets {
+					if cmp.Equal(targetExpected, foundTarget) {
+						break Outer
```

Comment:
Why not just `return`?

Reply:
Actually, this is supposed to `continue` (this is supposed to check the existence of each item in the array in the other array because Go lacks a `contains` method or any other basic operation). I refactored so this instead turns each array into a set and compares the set, allowing us to compare the entire list side by side instead of stopping at the first issue.

Reply:
I just ended up sorting one of the arrays by instance (since the other one is already sorted by that).